dev/doc/parsing.md

```diff
@@ -0,0 +1,397 @@
+
+When the parser is invoked, the call tells the parser which nonterminal to parse.  `vernac_control`
+is the start symbol for commands.  `tactic_mode` is the start symbol for tactics.
+Tactics give syntax errors if Coq is not in proof mode.  There are additional details
+not mentioned here.
+
+### Parsing productions
+
+Some thoughts, not to be taken as identifying all the issues:
+
+Since the parser examines only the next token to make a parsing decision (and perhaps
+because of other potentially fixable limitations), some productions have to be ordered
+or structured in a particular way to parse correctly in all cases.
+
+For example, consider these productions:
+
+  ```
+  command: [ [
+    | IDENT "Print"; p = printable -> { VernacPrint p }
+    | IDENT "Print"; qid = smart_global; l = OPT univ_name_list -> { VernacPrint (PrintName (qid,l)) }
+    | IDENT "Print"; IDENT "Module"; "Type"; qid = global ->
+        { VernacPrint (PrintModuleType qid) }
+    | IDENT "Print"; IDENT "Module"; qid = global ->
+        { VernacPrint (PrintModule qid) }
+    | IDENT "Print"; IDENT "Namespace" ; ns = dirpath ->
+        { VernacPrint (PrintNamespace ns) }
+    :
+
+  printable:
+    [ [ IDENT "Term"; qid = smart_global; l = OPT univ_name_list -> { PrintName (qid,l) }
+      | IDENT "All" -> { PrintFullContext }
+      | IDENT "Section"; s = global -> { PrintSectionContext s }
+      :
+  ```
+
+Reversing the order of the first two productions in `command` causes the `All` in `Print All` to
+be parsed incorrectly as a `smart_global`, making that command unavailable.  `Print Namespace nat.`
+still works correctly, though.
+
+Similarly, the production for `Print Module Type` has to appear before `Print Module <global>`
+in order to be reachable.
+
+Internally, the parser generates a tree that represents the possible prefixes for the
+productions of a nonterminal as described in
+[the Camlp5 documentation](http://camlp5.github.io/doc/htmlc/grammars.html#b:Rules-insertion).
+
+Here's another example in which the way the productions are written matters.  `OPT` at
+the beginning of a production doesn't always work well:
+
+  ```
+  command: [ [
+      | IDENT "Foo"; n = natural -> { VernacBack 1 }
+      | OPT (IDENT "ZZ"); IDENT "Foo" -> { VernacBack 1 }
+      :
+  ```
+
+`Foo.` looks like it should be accepted, but it gives a parse error:
+
+  ```
+  Unnamed_thm < Foo.
+  Toplevel input, characters 3-4:
+  > Foo.
+  >    ^
+  Error:
+  Syntax error: [prim:natural] expected after 'Foo' (in [vernac:command]).
+  ```
+
+Reversing the order of the productions doesn't help, but splitting
+the 'OPT' production into 2 productions works:
+
+  ```
+      | IDENT "Foo" -> { VernacBack 1 }
+      | IDENT "ZZ"; IDENT "Foo" -> { VernacBack 1 }
+      | IDENT "Foo"; n = natural -> { VernacBack 1 }
+
+  ```
+
+On the other hand, `OPT` works just fine when the parser has already found the
+right production.  For example `Back` and `Back <natural>` can be combined using
```

Comment:
I'm pretty sure we could improve the camlp5 engine to automatically do this expansion, i.e. at the time of inserting the rule in the tree, or even at runtime (at least when in the same entry).

Reply:
I'm not sure how often we need or would want to handle this case.  I included it in this write up because it was a simple example, not because I decided (or thought much about whether) the limitation really impacts users much.  I think I'd first analyze the grammar, including the numerous edits that `doc_grammar` makes for the doc and see how often this feature would be helpful.  I'd also consider whether such a change would impact other improvements we  may want to make to the parser--which requires having a wish list.

In a lot of the Coq code, we should think about how to make things simpler, more consistent and easier to understand/modify.  And add new features when they give a clear user benefit.